### PR TITLE
Added GPU implementation for resize nearest neighbor.

### PR DESCRIPTION
**Pull request by panmari**
Also added appropriate tests/benchmarks. Implementation for backwards step is still missing and will follow in another PR. I tried to model it after the code used for maxpooling.

### CODE REVIEW

**Comment by panmari** [comment]
Rebased again on master due to changes in `tensorflow/python/ops/image_ops_test.py`

**Comment by panmari** [comment]
@martinwicke could I please get some feedback on this?

**Comment by martinwicke** [comment]
Jenkins, test this please.

File: tensorflow/core/kernels/resize_nearest_neighbor_op.cc [hunk 1]
```diff
@@ -182,4 +186,65 @@ TF_CALL_REAL_NUMBER_TYPES(REGISTER_KERNEL);
...
+    OP_REQUIRES(context, sizes(0) > 0 && sizes(1) > 0,
+                errors::InvalidArgument("shape_t's elements must be positive"));
+
+    // Initialize shape to the batch size of the input, then add
+    // the rest of the dimensions
+    Tensor* output = nullptr;
+    OP_REQUIRES_OK(context, context->allocate_output(
+                                0, TensorShape({input.dim_size(0), sizes(0),
+                                                sizes(1), input.dim_size(3)}),
+                                &output));
```
Comment (martinwicke) [inline tensorflow/core/kernels/resize_nearest_neighbor_op.cc]: Can you indent this to 4 below the indent of OP_REQUIRES_OK?

**Comment by martinwicke** [comment]
Assuming the tests pass, LGTM.

**Comment by martinwicke** [comment]
That is, after the line comment.

File: tensorflow/core/kernels/resize_nearest_neighbor_op.cc (hunk 1 above)
Comment (panmari) [inline tensorflow/core/kernels/resize_nearest_neighbor_op.cc]: Done.

**Comment by vrv** [comment]
Cool, can you squash your commits?  We'll use this change as a way to test our eventual GPU test slave, maybe.

**Comment by panmari** [comment]
@vrv rebased on master and squashed, as requested.

**Comment by martinwicke** [comment]
Thanks! I'll wait to test until our GPU test machine comes online to test this. Should be today, I hope.

**Comment by martinwicke** [comment]
Sorry this is taking so long. Hardware is hard.

**Comment by panmari** [comment]
:computer: :hammer: ?

 :smile:

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by panmari** [comment]
Rebased on master again.

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by vrv** [comment]
doh, port.h was removed -- @panmari you'll probably have to update the code one more time :(

**Comment by martinwicke** [comment]
On the plus side, we finally have working GPU tests now.

**Comment by panmari** [comment]
Sure, will do! I'd have to adapt the code anyway to support the new flag
introduced recently...

**Comment by panmari** [comment]
I added another commit to support the newly added `align_cornes` flag and adapted some other things necessary. Unfortunately, on my local machine one of the tests in `//tensorflow/python:image_ops_test` is now failing. Could this be due to changes in the GPU allocator?

**Comment by vrv** [comment]
What's the failure in particular?

**Comment by panmari** [comment]
@vrv in the test where I compare the cpu resizing with the gpu resizing, I get completely different results there even though the other tests pass

```
FAIL: testCompareNearestNeighbor (__main__.ResizeImagesTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/panmari/.cache/bazel/_bazel_panmari/52e003c51984ee00b8d04ed6ebb99872/tensorflow/bazel-out/local_linux-opt/bin/tensorflow/python/image_ops_test.runfiles/tensorflow/python/ops/image_ops_test.py", line 990, in testCompareNearestNeighbor
    self.assertAllClose(cpu_val, gpu_val, rtol=1e-5, atol=1e-5)
  File "/home/panmari/.cache/bazel/_bazel_panmari/52e003c51984ee00b8d04ed6ebb99872/tensorflow/bazel-out/local_linux-opt/bin/tensorflow/python/image_ops_test.runfiles/tensorflow/python/framework/test_util.py", line 435, in assertAllClose
    np.testing.assert_allclose(a, b, rtol=rtol, atol=atol)
  File "/usr/local/lib/python2.7/dist-packages/numpy/testing/utils.py", line 1359, in assert_allclose
    verbose=verbose, header=header)
  File "/usr/local/lib/python2.7/dist-packages/numpy/testing/utils.py", line 713, in assert_array_compare
    raise AssertionError(msg)
AssertionError: 
Not equal to tolerance rtol=1e-05, atol=1e-05

(mismatch 100.0%)
 x: array([[[[  0.,   1.,   2.],
         [  0.,   1.,   2.],
         [  3.,   4.,   5.],...
 y: array([[[[  5.903004e+20,   1.406543e+01,   4.097450e-34],
         [  5.903004e+20,   1.406543e+01,   4.097450e-34],
         [  5.903004e+20,   1.406543e+01,   4.097450e-34],...

----------------------------------------------------------------------
Ran 16 tests in 0.759s

FAILED (failures=1)
   1.50654290e-04   9.99999975e-06
   5.90300359e+15   1.50654290e-04   9.99999975e-06   5.90300359e+15
   1.50654290e-04   9.99999975e-06   5.90300359e+15   1.50654290e-04
   9.99999975e-06   5.90300359e+15   1.50654290e-04   9.99999975e-06
   5.90300359e+15   1.50654290e-04   9.99999975e-06   5.90300359e+15
   1.50654290e-04   9.99999975e-06   5.90300359e+15   1.50654290e-04
   9.99999975e-06   5.90300359e+15   1.50654290e-04   9.99999975e-06
   5.90300359e+15   1.50654290e-04   9.99999975e-06   5.90300359e+15
   1.50654290e-04   9.99999975e-06   5.90300359e+15   1.50654290e-04
   9.99999975e-06   5.90300359e+15   1.50654290e-04   9.99999975e-06]
area_ratio_hist  [ 70  73  90 100 133 154 162  95  80  43]
```

**Comment by vrv** [comment]
@panmari: isn't that specifically the kernel you added that is showing different behavior between CPU and GPU?  That probably needs to be fixed, unless I'm mis-reading the error.

**Comment by panmari** [comment]
@vrv sorry, it took some time until I got around fixing this. Seemed like `np.float32` was interpreted differently due to some change since 0.6, leading to the failing test. I generalized my code a bit to make an extension to other types more easy.

**Comment by vrv** [comment]
@tensorflow-jenkins: test this please

(i swear, we'll get this in!)

**Comment by vrv** [comment]
Merged, woohoo!

**Comment by panmari** [comment]
Cool, glad it finally happened :smile:

**Pull request closed without merging**